### PR TITLE
Pagination (+ display composers with mock data)

### PR DESCRIPTION
**Pull request by lustoykov**
Still a bit rough, but creating the pull request before it got too big.

### CODE REVIEW

File: react/src/components/composers-app/Composers.jsx [hunk 1]
```diff
@@ -0,0 +1,97 @@
...
+      showSuggestions: false,
+      query: '',
+    };
+  }
+
+  componentWillMount() {
+    loadComposers().then(result => {
+      const { composers, composersTotalCount } = result;
+      this.setState({ composers, composersTotalCount, currentPage: 1 });
+    });
```
Comment (lustoykov) [inline react/src/components/composers-app/Composers.jsx]: remove code and call `this.onSearchClick()` directly on `componentWillMount`

File: react/src/components/composers-app/utils/loadComposers.js [hunk 2]
```diff
@@ -0,0 +1,18 @@
...
+    const begin = (page - 1) * 15;
+    const end = begin + limit;
+    const filteredComposers = composers.filter(entity => entity.title.toLowerCase().includes(query.toLowerCase()));
+
+    resolve({
+      composers: filteredComposers.slice(begin, end),
+      composersTotalCount: filteredComposers.length,
+    });
+  });
+}
```
Comment (lustoykov) [inline react/src/components/composers-app/utils/loadComposers.js]: Simulates request (returns Promise with composers and composersTotalCount)

File: react/src/components/composers-app/utils/loadSuggestions.js [hunk 3]
```diff
@@ -0,0 +1,11 @@
...
+  { title: 'Compos' },
+  { title: 'Composer' },
+  { title: 'Composer1' },
+];
+
+export default function loadSuggestions(query = '') { // eslint-disable-line no-unused-vars
+  return Promise.resolve({
+    suggestions,
+  });
+}
```
Comment (lustoykov) [inline react/src/components/composers-app/utils/loadSuggestions.js]: Simulates request (returns Promise with suggestions)

File: react/src/components/pagination/PaginationSearch.jsx [hunk 4]
```diff
@@ -0,0 +1,20 @@
+import React from 'react';
+import { InputGroup, InputGroupAddon, Input } from 'reactstrap';
+
+
+const PaginationSearch = props => ( // eslint-disable-line no-unused-vars
```
Comment (lustoykov) [inline react/src/components/pagination/PaginationSearch.jsx]: remove comment

**Review by lustoykov — COMMENTED** [review]
PS. Just noticed a small bug, where the page number (1,2,3...7) is not hovered correctly (I have to save that in state), but data is loaded with correctly clicked page.

File: react/webpack.config.js [hunk 5]
```diff
@@ -57,6 +59,10 @@ module.exports = {
       template: 'index.html',
       inject: 'body',
       chunks: ['personGraph'],
+      filename: 'composers-app.html',
```
Comment (chaoran-chen) [inline react/webpack.config.js]: Here is a merge mistake ;)

File: react/src/components/pagination/PaginationContent.jsx [hunk 6]
```diff
@@ -0,0 +1,33 @@
...
+  const {
+    EntityComponent,
+    entities,
+    className,
+  } = props;
+
+  return (
+    <div className={`${className}`}>
+      <ListGroup>
+        {_.map(entities, (entity, index) => (
```
Comment (chaoran-chen) [inline react/src/components/pagination/PaginationContent.jsx]: Why are you using lodash instead of Array.prototype.map?

**Comment by chaoran-chen** [comment]
I get an error when I open /composers-app.html, have I forgotten something?
error

**Comment by lustoykov** [comment]
@chaoran-chen `yarn install`?

File: react/src/components/pagination/PaginationContent.jsx (hunk 6 above)
Comment (lustoykov) [reply react/src/components/pagination/PaginationContent.jsx]: They behave slightly differently, not that it makes difference in this case. Also lodash's methos are usually faster and is more consistent with rest of my code

tl;dr: for no reason 👍 

http://jsperf.com/array-map-vs-lo-dash-map

**Comment by chaoran-chen** [comment]
I did.

**Comment by lustoykov** [comment]
@chaoran-chen can you try again? there were a lot of conflicts when I rebased, hope it's OK.

**Comment by lustoykov** [comment]
[https://drive.google.com/open?id=0B_SI7o_qKKH2ME03WE1vOVQtcGM](url)
[https://drive.google.com/open?id=0B_SI7o_qKKH2MndqR3dOejlINmc](url)
[https://drive.google.com/open?id=0B_SI7o_qKKH2QzZ3Szcyd1ZuNFE](url)

**Comment by lustoykov** [comment]
Image of Pagination

**Comment by lustoykov** [comment]
[url=https://postimg.org/image/fig989bgt/][img]https://s22.postimg.org/fig989bgt/2_display_content.png[/img][/url]

[url=https://postimg.org/image/8t9ps8q4t/][img]https://s22.postimg.org/8t9ps8q4t/3_loading_suggestions.png[/img][/url]

[url=https://postimg.org/image/l96flzjgt/][img]https://s22.postimg.org/l96flzjgt/5_click_suggestion.png[/img][/url]

**Comment by lustoykov** [comment]
GitHub Logo
GitHub Logo
GitHub Logo

**Comment by lustoykov** [comment]
# We'll add more data and integrate works, etc once the API for that is done. For now only display the title of the composer.

2_display_content
3_loading_suggestions
6_display_content

**Comment by lustoykov** [comment]
We'll add more data and integrate works, etc once the API for that is done. For now only display the title of the composer.

File: react/src/components/pagination/PaginationSearch.jsx (hunk 4 above)
Comment (chaoran-chen) [reply react/src/components/pagination/PaginationSearch.jsx]: Dont't forget to remove it ;)

**Review by chaoran-chen — APPROVED** [review]
It works now :)
A small feature suggestion for the next iteration: please let Enter trigger the search action in the search bar.

**Comment by lustoykov** [comment]
Thank you @chaoran-chen great suggestion.

File: react/scss/composers-app.scss [hunk 7]
```diff
@@ -0,0 +1,31 @@
+@import '~bootstrap/dist/css/bootstrap.css';
+
+.pagination__header, .pagination__search, .pagination__suggestions, .pagination__content, .pagination__footer {
+  margin: 20px 0 20px 0;
+}
+
+.pagination-footer__separator {
```
Comment (martomi) [inline react/scss/composers-app.scss:7]: I'm still quite new to the BEM style. Could you elaborate what is the difference / advantage of `pagination-footer__separator` vs `pagination__footer__separator` as a class name? Thanks.

Comment (lustoykov) [reply react/scss/composers-app.scss:7]: pagination-footer__separator is a subcomponent __separator that belongs to pagination-footer. pagination-footer is the main component.

pagination__footer__separator would mean that there is a component __footer that has entity __separator IN pagination component. (but it is actually a __separator in pagination-footer component). If we did pagination__footer__separator, then the pagination-footer would need to assume that it belongs to pagination component and I don't like assumptions. If they are in different files, I separate them as own entities. pagination__footer__separator couples pagination on footer and separator. pagination-footer__separator couples separator on pagination-footer. 

The fact that I needed 10 lines to explain it means: it's a bit in the grey area, but I feel like there is a better separation of concerns this way.
